test/integration/integration_test.cc

```diff
@@ -283,6 +283,39 @@ TEST_P(IntegrationTest, HittingGrpcFilterLimitBufferingHeaders) {
               HeaderValueOf(Headers::get().GrpcStatus, "2")); // Unknown gRPC error
 }
 
+TEST_P(IntegrationTest, TrailersWithHttp1AndHttp2) {
+  setUpstreamProtocol(FakeHttpConnection::Type::HTTP2);
+  Http::TestHeaderMapImpl response_trailers{{"response1", "trailer1"}, {"response2", "trailer2"}};
+
+  initialize();
+  codec_client_ = makeHttpConnection(lookupPort("http"));
+  auto encoder_decoder =
+      codec_client_->startRequest(Http::TestHeaderMapImpl{{":method", "POST"},
+                                                          {":path", "/test/long/url"},
+                                                          {":scheme", "http"},
+                                                          {":authority", "host"},
+                                                          {"TE", "trailers"},
+                                                          {"Accept", "*/*"}});
+  request_encoder_ = &encoder_decoder.first;
+  codec_client_->sendData(*request_encoder_, 0, true);
+
+  auto response = std::move(encoder_decoder.second);
+
+  waitForNextUpstreamRequest();
+  upstream_request_->encodeHeaders(default_response_headers_, false);
+  upstream_request_->encodeData(5, false);
+  upstream_request_->encodeTrailers(response_trailers);
+  response->waitForEndStream();
+
+  ASSERT_TRUE(upstream_request_->complete());
+  ASSERT_TRUE(response->complete());
+
+  EXPECT_EQ(5, response->body().size());
+  EXPECT_EQ("aaaaa", response->body());
+  EXPECT_EQ("200", response->headers().Status()->value().getStringView());
+  EXPECT_THAT(*response->trailers(), HeaderMapEqualRef(&response_trailers));
```

Comment:
@htuch Would like your input on this since I have been spinning my wheels on this for a while...

Trailers are null here, which fails the test.

However what is puzzling is that if I do a curl against the integration test itself, I get the trailers back

![image](https://user-images.githubusercontent.com/5170710/67526525-d4945280-f669-11e9-8948-b361f43b2a50.png)

So I must be setting up this test completely wrong here if I can see the trailers when curling the test itself.

Reply:
+ @wlhee

Reply:
The test is using the Envoy product H1 codec to produce the trailers. Do you have the correct plumbing on the decode path as well as encode?

Reply:
Yea working on that now. I realized I needed to change the onHeaders callbacks in `codec_impl` to not ignore trailers. Though I'm still confused as to how the above was showing them? Perhaps the trailers above are actually part of the body?

Reply:
When you `curl`, it will be the `curl` client that is doing the decode..

Reply:
Ahhhh okay. That makes a lot more sense now!